### PR TITLE
Add specific errors for sealing failures

### PR DESCRIPTION
**Pull request by nick-mobilecoin**
Add errors that provide more information on why a failure to seal data
occurs.

### CODE REVIEW

**Comment by nick-mobilecoin** [comment]
Current dependencies on/for this PR:
* main
  * **PR #158** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx/158" target="_blank">Graphite</a> 
    * **PR #159** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx/159" target="_blank">Graphite</a> 
      * **PR #163** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx/163" target="_blank">Graphite</a> 
        * **PR #164** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx/164" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx/164?utm_source=stack-comment).

**Comment by nick-mobilecoin** [comment]
This initial change only updates the _current_ error handling in sealing to use a seal specific error.
Subsequent commits will provide _mirror_ parameter validation that `sgx_seal_data_ex` performs.

File: tservice/src/seal.rs
Comment (nick-mobilecoin) [inline tservice/src/seal.rs:39]: Up for suggestions on these names, I may be copying a bit too much from the SGX API.
My thought was instead of having an error per each failure have an error _group_ and provide a message with the details

File: tservice/src/seal.rs
Comment (nick-mobilecoin) [inline tservice/src/seal.rs:34]: Up for ideas on this, it feels like a misnomer in the SGX SDK, but I think it's nice to call out the function name.

**Review by NotGyro — APPROVED** [review]
Looks good to me - I can see ways that it could be more specific, but having these broader categories of errors could be much easier to work with.

**Review by jcape — CHANGES_REQUESTED** [review]
I think these errors are pointing in the wrong direction, unfortunately. The real issue that needs solving is "give the caller a more descriptive/fine-grained error code when the SDK would return something obtuse like `INVALID_PARAMETER`."

Like, INVALID_PARAMETER could mean "buffer not in enclave" or "buffer length too short" or "plaintext + mac length > u32::MAX" or any number of things, I'm saying you shouldn't have to read the SGX FFI PDF to figure that out, it should be immediately evident, even if that means duplicating the input validation that the SDK is doing in our rust code.

File: tservice/src/seal.rs
Comment (nick-mobilecoin) [reply tservice/src/seal.rs]: second attempt, up

File: tservice/src/seal.rs
Comment (jcape) [inline tservice/src/seal.rs:32]:
```suggestion
    /** The combined plaintext ({data_size}) and authenticated data
     *  ({aad_size}) sizes are larger than 4GiB */
```

File: tservice/src/seal.rs
Comment (jcape) [inline tservice/src/seal.rs:35]:
```suggestion
    /** The destination buffer must be at least {needed_size} bytes,
     *  {buffer_size} was given */
```

**Review by jcape — APPROVED** [review]
LGTM, some minor rewording of the `Display` format strings. For the logging case, it's generally prefered to keep them as a single sentence with no errors, due to how nested errors are displayed. An error hierarchy like:

`FooError::Frobnicate(FrobnicateError::BlahBityBlah(BlahError::WastefulAlloc)`

will end up getting printed as:

`"Foo error while frobnicating: Frobincating requires blahbityblah: Blah is larger than necessary."`

File: tservice/src/seal.rs
Comment (nick-mobilecoin) [reply tservice/src/seal.rs]: error removed now

**Comment by nick-mobilecoin** [comment]
**Graphite Merge Job**

Current status: 🔄 Waiting for checks to pass

This pull request is currently waiting for CI and mergeability checks to pass before being merged.

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx/164?utm_source=stack-comment).


Job Reference: MYnB2IroAk46QVSy8SQp

**Pull request merged**
